csrc/dynamic_transform.h

```diff
@@ -0,0 +1,76 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#pragma once
+
+#include <c10/macros/Export.h>
+
+#include <ir_all_nodes.h>
+#include <iter_visitor.h>
+#include <transform_view.h>
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+namespace nvfuser {
+
+class Fusion;
+class ExpressionEvaluator;
+class DynamicTransformInfoBuilder;
+
+class TORCH_CUDA_CU_API DynamicTransformInfo {
+ public:
+  bool operator==(const DynamicTransformInfo& other) const;
+
+  bool operator!=(const DynamicTransformInfo& other) const {
+    return !(*this == other);
+  }
+
+  std::string toString() const;
+
+  DynamicTransformInfo(Fusion* fusion) : fusion_(fusion) {}
+
+  Fusion* fusion() const {
+    return fusion_;
+  }
+
+  const std::vector<std::pair<TensorView*, AnalyzeViewResult>>
+  getReshapeTransforms() const {
+    return reshape_transforms_;
+  }
+
+  // TODO: Make it private
+ public:
+  Fusion* fusion_ = nullptr;
+  std::vector<std::pair<TensorView*, AnalyzeViewResult>> reshape_transforms_;
+
+  // TODO: resize transforms
+
+  friend class DynamicTransformInfoBuilder;
+};
+
+class TORCH_CUDA_CU_API DynamicTransform {
+ public:
+  static DynamicTransformInfo getConcretizationInfo(
+      Fusion*,
+      ExpressionEvaluator* expr_eval);
+
+  static void concretizeFusion(Fusion*, const DynamicTransformInfo& info);
```

Comment:
Looks like concretizeFusion does modify the object pointed by `Fusion*` since it's not a `const Fusion*`?

cc'ing @kevinstephano since this likely would affect how/where `Fusion*` is cached on the python side.

Reply:
Yes, that's what the current design does. We could change if that's more preferred. For example, we could just create a copy, concretize it and return the new copy. 

I think this is at this moment just an interface design question. We may want to have a different interface, but it seems still too early to think about concrete interfaces.